### DISCUSSION TITLE
Regarding Architecture & Technologies

### DISCUSSION THREAD

**Discussion by jonasdacruz** [jonasdacruz/mars-coms#2] [General]
Main idea of this project is to check the current possibility of deploying a full cloud solution using free tiers of different providers, but avoiding vendor lock in.

My first proposal is the following:
- Use expo.dev for building, distributing and testing React Native mobile apps for Android and iOS
- Use serverless.com framework to deploy functions written in Go in several providers such as AWS and Google
- Use cloudamqp.com to queue different tasks and delegate part of the complexity
- Use mongodb.com/atlas to persist data

Open questions for me are:
- what to use in order to store files? MongoDB has the possibility for small files it looks fine, however I'm not sure is the right answer.
- what to use for monitoring the whole system

**Comment by alejandrogr** [comment]
![overview](https://user-images.githubusercontent.com/701346/176412642-30867be5-9048-4caa-988b-734f57ee54bb.png)

Having that simple diagram in mind:

- expo.dev seems like a good fit for the lifecyle of the the front app. I've never used it but it seems like a powerful option
- serveless framework works like a charm for the serveless part. I have used it in the past always with good results
- cloudamqp sounds also good to handle the message queue. Keep in mind that we need that the event broker has support for [delayed messages](https://blog.rabbitmq.com/posts/2015/04/scheduling-messages-with-rabbitmq) 
- MongoDB Atlas for the database is ok
- For the notification service I will go with onesignal.com Their free tier will cover everything we need.

For file storage I will not go for GridFS inside mongo. I think that we will need something more versatile (a proper filestore like S3 or azure blob or similar).

**Comment by alejandrogr** [comment]
![overview_functional](https://user-images.githubusercontent.com/701346/176660128-00bc4f76-dc3d-4c7e-85ba-ab2ce373516f.png)

To help us start visualizing things :D

# Overview Functional Diagram

Actors:
- Sender: the user sending a message
- Receiver: the user receiving a message
- App: the MarsComs application
- Message Publisher: service to publish/schedule messages
- Message Queue: an event broker to store messages
- Message Deliverer: service to deliver messages
- Notification Service: service for push notifications
- Database: store all information needed for MarsComs to work

## Functional Descriptions


### An user SignUp

- The SignUp flow is handled by the **App** including the user validation (phone validation, store info in **Database**...)
- The user device is subscribed to push notifications in the **Notification Service**

### An user sends a message

- The **Sender** communicates with the **Message Publisher** service to send a message
- The **Message Publisher** schedule the message in the **Message Queue** and store the message in the database
- The **Sender** can now track the messages status and timeline through the **App**
- When the time comes the **Message Queue** deliver the message to the **Message Deliverer**
- The **Message Deliverer** updates the message status info in the **Database** and calls the **Notifaction Service**
- The **Notification Service** sends a push notification to the **Receiver**
